### ISSUE TITLE
Socket backlog setting in Kitura

### ISSUE THREAD

**Issue by simplepanda** [IBM-Swift/Kitura#589]
Hi all,

Investigating some benchmark issues while implementing FastCGI in Kitura and I kept running into an annoying problem where my web server (nginx) would return about 10% of it's responses as error 502 (bad gateway), claiming that the upstream (Kitura) disconnected prematurely.

I instrumented my code like crazy and couldn't find a problem. It seems that my code was never even seeing the bad connections in the first place...

Running a test like this:

```
wrk -t4 -c150 -d30s http://127.0.0.1
```

I eventually figured out that:

Errors reported by WRK = Errors reported by Nginx
Total attempts by WRK - Errors reported by WRK = total attempts received by Kitura.

So this led me into BlueSocket to see what was up, where I stumbled on this little beauty in Socket.swift:

```
public static let SOCKET_DEFAULT_MAX_CONNECTIONS		= 5
```

Not exactly named well (though it's not named well in traditional BSD man pages either, but I digress), this variable is eventually used here:

```
public var maxPendingConnections: Int = Socket.SOCKET_DEFAULT_MAX_CONNECTIONS
```

Which ultimately ends up here:

```
if Darwin.listen(self.socketfd, Int32(maxPendingConnections)) < 0 {
    throw Error(code: Socket.SOCKET_ERR_LISTEN_FAILED, reason: self.lastError())
}
```

So this isn't so much a "Max Connections" as it is the TCP socket backlog parameter. 5 isn't really a great value for server applications under high concurrency (I adjusted this up to OS X's max backlog value (128), per:

```
$ sysctl kern.ipc.somaxconn
```

... and all my benchmarking errors immediately disappeared. 0 failures from Nginx (no 502 errors) and zero direct benchmarking failures (mostly socket connect errors) to Kitura's web server.

Suggesting a patch to Kitura (more likely Kitura-Net) to update the listen backlog.

Nginx uses the following (where "other platforms" basically means Linux):

> backlog=number sets the backlog parameter in the listen() call that limits the maximum length for the queue of pending connections. By default, backlog is set to -1 on FreeBSD, DragonFly BSD, and Mac OS X, and to 511 on other platforms.

-1 should mean "maximum allowable" on OS X.

Thoughts?

**Comment by simplepanda** [comment]
In the FastCGI implementation I'm working on, I've implemented it this way:

```
self.listenSocket = try Socket.create()
            
#if os(Linux)
self.listenSocket?.maxPendingConnections = 511
#else
self.listenSocket?.maxPendingConnections = -1
#endif
```

This happens before listen() begins and it solves the backlog problems nicely for FastCGI. Could also be applied to HTTP.

**Comment by ianpartridge** [comment]
@billabt ?

**Comment by shmuelk** [comment]
In an experimental branch of Kitura that supports keep-alive on Linux I have raised the connection backlog to 100, in Kitura.

We could add that code to the main branch of Kitura-net. We could even use the logic that @simplepanda found in Nginx. Although I'd like to better understand the ramification o having a high connection backlog. I think the use case of Kitura is more towards micro-services and less towards a full stand up web server as Nginx is.

**Comment by billabt** [comment]
I'm on vacation this week but will take a look when I get the chance...

**Comment by buckstephenh** [comment]
For what it's worth, in high concurrency applications, which are often never imagined by the developer,  which includes microservices and anything listening on tcp, when there is a slow down behind kitura, for example, an intermittently slow sql response, it is generally cheaper and more stable to leave the connections in the accept backlog rather than consuming thread resources and trying to over-parallelize and take down the whole show, or cause thread pool growth which keeps things slow after the intermittent problem resolves.   So I agree with @simplepanda recommendations.

**Comment by collinhundley** [comment]
I'm experiencing this same issue occasionally - Kitura refuses connections from Nginx so it returns `502`. I wouldn't say it's 10% of the time (probably closer to 1-2% of the time), but in my case it rarely receives concurrent connections.

**Comment by simplepanda** [comment]
@collinhundley Are you using a release or the master branch FastCGI support?

**Comment by shmuelk** [comment]
A PR of mine that adds keep alive support on both Linux and OSX is being reviewed and changes are being made based on the reviews.

Once that PR is merged, this problem should go away.

**Comment by shmuelk** [comment]
The keep alive work (issue IBM-Swift/Kitura#435) has been merged into master and is available as Kitura 0.23.x and Kitura-net 0.23.x.

The performance on Linux has improved nicely and on OSX extremely well.

**Comment by shmuelk** [comment]
Is this issue still happening or has the combination of HTTP keep alive, raising the socket backlog (t o one hundred), and other performance improvements, solved the problem?

**Comment by ianpartridge** [comment]
I'm closing this as fixed by the Keep-Alive work.
